File: src/ccycles/scene_parameters.cpp
/**
Copyright 2014-2017 Robert McNeel and Associates

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
**/

#include "internal_types.h"

#define SCENE_PARAM_BOOL(scene_params_id, varname) \
	PARAM_BOOL(scene_params, scene_params_id, varname)

#define SCENE_PARAM_CAST(scene_params_id, typecast, varname) \
	PARAM_CAST(scene_params, scene_params_id, typecast, varname)

std::vector<ccl::SceneParams*> scene_params;

/* Create scene parameters, to be used when creating a new scene. */
unsigned int cycles_scene_params_create(
	unsigned int shadingsystem, unsigned int bvh_type, 
	unsigned int use_bvh_spatial_split, 
	int bvh_layout, unsigned int persistent_data)
{
	ccl::SceneParams* params = new ccl::SceneParams();

	params->shadingsystem = (ccl::ShadingSystem)shadingsystem;
	params->bvh_type = (ccl::BVHType)bvh_type;
	params->use_bvh_spatial_split = use_bvh_spatial_split == 1;
	params->bvh_layout = (ccl::BVHLayout)bvh_layout;
    // TODO: XXXX check if persistent data is done differently
	//params->persistent_data = persistent_data == 1;

	scene_params.push_back(params);

    /*
	logger.logit("Created scene parameters ", scene_params.size() - 1, "\n\tshading system: ", params->shadingsystem, "\n\tbvh_type: ", params->bvh_type, "\n\tuse_bvh_spatial_split: ", params->use_bvh_spatial_split, "\n\tuse_qbvh: ", params->bvh_layout, "\n\tpersistent data: ", params->persistent_data);*/

	return (unsigned int)(scene_params.size() - 1);
}

/* Set scene parameters*/
void cycles_scene_params_set_bvh_type(unsigned int scene_params_id, unsigned int bvh_type)
{
	SCENE_PARAM_CAST(scene_params_id, ccl::BVHType, bvh_type)
}

void cycles_scene_params_set_bvh_spatial_split(unsigned int scene_params_id, unsigned int use_bvh_spatial_split)
{
	SCENE_PARAM_BOOL(scene_params_id, use_bvh_spatial_split)
}
void cycles_scene_params_set_qbvh(unsigned int scene_params_id, unsigned int use_qbvh)
{
    // TODO: XXXX revisit BVH settings, there are different ones now.
    // For now default to BVH_LAYOUT_BVH2, but there will be much better ones
	ccl::BVHLayout bvh_layout = ccl::BVHLayout::BVH_LAYOUT_BVH2;
	SCENE_PARAM_CAST(scene_params_id, ccl::BVHLayout, bvh_layout)
}

void cycles_scene_params_set_shadingsystem(unsigned int scene_params_id, unsigned int shadingsystem)
{
	SCENE_PARAM_CAST(scene_params_id, ccl::ShadingSystem, shadingsystem)
}
void cycles_scene_params_set_persistent_data(unsigned int scene_params_id, unsigned int persistent_data)
{
    // TODO: XXXX no longer exists
	//SCENE_PARAM_BOOL(scene_params_id, persistent_data)
}

File: csycles/SceneParameters.cs
/**
Copyright 2014-2024 Robert McNeel and Associates

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
**/

namespace ccl
{
	/// <summary>
	/// Wrapper for creating scene parameters.
	/// 
	/// Note that this is only used for *setting*, not fetching parameters.
	/// 
	/// BVH stands for Bound Volume Hierarchy (https://en.wikipedia.org/wiki/Bounding_volume_hierarchy)
	/// </summary>
	public class SceneParameters
	{
		/// <summary>
		/// Get the Id for these scene parameters.
		/// </summary>
		public uint Id { get; }

		private Client Client { get; }

		/// <summary>
		/// Create scene parameters object with given parameters
		/// </summary>
		/// <param name="client">Reference to the client</param>
		/// <param name="shadingSystem">ShadingSystem to use</param>
		/// <param name="bvhType">BvhType to use (dynamic or static)</param>
		/// <param name="bvhCache">True if BVH should be cached</param>
		/// <param name="bvhSpatialSplit">True if BVH spatial splits should be used</param>
		/// <param name="bvhLayout">BvhLayout to use. Default means Bvh8</param>
		/// <param name="persistentData">True if data should be persisted</param>
		public SceneParameters(Client client, ShadingSystem shadingSystem, BvhType bvhType, bool bvhSpatialSplit, BvhLayout bvhLayout, bool persistentData)
		{
			Client = client;
			Id = CSycles.scene_params_create(shadingSystem, bvhType, bvhSpatialSplit, bvhLayout, persistentData);
		}

		/// <summary>
		/// Set the <c>BvhType</c> to be used
		/// </summary>
		public BvhType BvhType
		{
			set
			{
				CSycles.scene_params_set_bvh_type(Id, value);
			}
		}

		/// <summary>
		/// Set to true if BVH should use spatial splits.
		/// </summary>
		public bool BvhSpatialSplit
		{
			set
			{
				CSycles.scene_params_set_bvh_spatial_split(Id, value);
			}
		}

		/// <summary>
		/// Set to true if QBVH should be used.
		/// </summary>
		public bool QBvh
		{
			set
			{
				CSycles.scene_params_set_qbvh(Id, value);
			}
		}

		/// <summary>
		/// Set the <c>ShadingSystem</c> to use.
		/// 
		/// Note that currently only <c>ShadingSystem.SVM</c> is supported.
		/// </summary>
		public ShadingSystem ShadingSystem
		{
			set
			{
				CSycles.scene_params_set_shadingsystem(Id, value);
			}
		}

		/// <summary>
		/// Set to true if BVH should be persisted between frames.
		/// </summary>
		public bool PersistentData
		{
			set
			{
				CSycles.scene_params_set_persistent_data(Id, value);
			}
		}
	}
}
